blog/2021/02/julia-1.6-highlights.md

```diff
@@ -0,0 +1,390 @@
+
+Earlier versions of Julia were somewhat conservative, and invalidated old code
+in some circumstances where there was no actual change in dispatch. Moreover,
+there were many places where Julia and its standard libraries were written in a
+way that defeated Julia’s type-inference. Because the compiler sometimes had to
+invalidate code just because a new method _might_ apply, any uncertainty about
+types magnifies the risk and frequency of invalidation. In older versions of
+Julia, the combination of these effects made invalidation widespread: just
+loading certain packages led to invalidation of up to 10% of Julia’s precompiled
+code. The delay for recompilation could sometimes make interactive sessions feel
+sluggish. When invalidation occurred in Julia’s package-loading code, it also
+delayed loading of the next package, contributing to long waits for using
+SomePkg when SomePkg depends on other packages.
+
+In 1.6, the scheme for invalidating old code has been made more accurate and
+selective. Moreover, Julia and its standard libraries received a thorough
+makeover to help type inference arrive at a concrete answer more often. The
+result is a leaner, faster Julia that is far more impervious to method
+invalidation, and feels considerably more responsive and nimble in interactive
+sessions. Related blog post: [Analyzing sources of compiler latency in Julia:
+method invalidations](https://julialang.org/blog/2020/08/invalidations/).
+
+## Compiler latency reduction
+
+_Jameson Nash and Jeff Bezanson_
+
+Load times and compilation pauses continue to be a significant nuissance for
+Julia users, and one of our main technical challenges. We try to make at least
+a little progress on it in every release. This time, there aren't any major
+breakthroughs, but we do have some modest improvements due to work on the
+method table data structure.
+
+Method specificity is a partial order, and prior to 1.6 we stored methods in
+sorted order. We also attempted to identify ambiguous methods on insertion,
+hoping to avoid repeating the work for each future query.
+Unfortunately, sorting a partial order requires quadratic time, and this time
+began to show up prominently during package loading (when a package's methods
+need to be inserted into the currently-active method tables).
+
+We improved things by making the process lazier, moving sorting and ambiguity
+detection into the algorithm for finding matching methods.
+This algorithm runs very often, so it was not at first intuitive that this
+change would help.
+But the key is that the vast majority of queries are for specific enough
+types that most possible matches can be eliminated easily, leaving
+many fewer inputs to the most expensive steps.
+
+The main visible improvement here is to package loading, which is now often
+significantly faster. For example, `using Plots` goes from about 6 seconds to
+about 3 seconds for me (after precompilation).
```

Comment:
This 2x bundles together the advances due to method sorting with the fact that we don't invalidate `Base.require` or `Pkg` methods so often. See whether you like the partial description in the invalidations section and whether we can combine these usefully (I wrote it anticipating you'd have something like this here, but obviously didn't know exactly how it would come out).